### PR TITLE
Null string values adjustments, nullable enums support, accept Oid.Name as string

### PR DESCRIPTION
**Pull request by tchaloupka**
When string value is null it should be the same in the DB.
Same goes with Nullable!string and both types should be also read as null from the DB

### CODE REVIEW

**Comment by denizzzka** [comment]
Probably, when string value is `null` in D it is "" (empty string) in Postgres

**Comment by tchaloupka** [comment]
Ok, I tried to look in other language libs to be somehow consistent.
* **Go** - throws error on reading null to string: https://github.com/go-sql-driver/mysql/issues/34
* **Python** - returns None for NULL values: https://stackoverflow.com/questions/3086265/python-postgres#3086797
* **.Net DataReader** - throw Exception: https://stackoverflow.com/questions/1772025/sql-data-reader-handling-null-column-values#1772037
* **Rust** - throws error: https://stackoverflow.com/questions/48067493/how-do-i-check-if-a-column-is-null-using-rust-postgres

So I would probably rather throw an error too so we don't introduce some unwanted conversions here.
Same as ie reading null value as int.

What do you think?

**Comment by denizzzka** [comment]
In D we have:
```D
assert(cast(string) null == "");
```

I.e. `null` string value threated as empty string. So, it isn't need to special case for D's null strings because it is empty ("") strings, or string with length == 0.

> Go - throws error on reading null to string: go-sql-driver/mysql#34

Right behaviour, as I think. And it is already implemented in `dpq2` for all `Value`:
https://github.com/denizzzka/dpq2/blob/master/src/dpq2/value.d#L87

(I do not consider here about `Nullable` values.)

**Comment by tchaloupka** [comment]
Ok so I'll change this so null string will be saved as empty string and reading null as string will throw an error.

**Comment by denizzzka** [comment]
> and reading null as string will throw an error.

D's null string threated as empty string by D

**Comment by tchaloupka** [comment]
Ok in D this is true:
```D
string s = null;
assert(s is null);
assert(s == "");
```

But what should be the result of `assert(s is null)` for a value read from NULL DB value?
As I understand, you want it to be false? So null string will be written as empty and read as empty?
Because if we write null string as an empty string and we read it back so `is null` is true, it seems to be inconsistent to me.

**Comment by denizzzka** [comment]
> But what should be the result of assert(s is null) for a value read from NULL DB value?

It will be Value.isNull == true

If you want to create PG NULL Value use special ctor or null `Nullable` value

**Comment by tchaloupka** [comment]
We don't understand each other.
What should be result of:
```D
auto res = conn.exec(...);
string name = res[0][0].as!string;
bool b = name is null;
```

My point is that if we do this:
```D
string s = null;
auto v = s.toValue;
```
Then written to DB column `== ''`

Then we should read NULL values from DB so that `bool b = name is null` will return false otherwise it would be inconsistent as we can't write null string as an empty string and return null column as a null string.
Agree?

**Comment by denizzzka** [comment]
> What should be result of:
```D
auto res = conn.exec(...);
string name = res[0][0].as!string;
bool b = name is null;
```
b is unknown, it depends from result string length.

> My point is that if we do this:
```D
string s = null;
auto v = s.toValue;
```
>Then written to DB column == ''

Agreed.

> Then we should read NULL values from DB so that bool b = name is null will return false otherwise it would be inconsistent as we can't write null string as an empty string and return null column as a null string.
> Agree?

No. We can write D's `null` string as an empty (length == 0) `not NULL` string. (If it isn't works something is broken.)
And we can return `isNull` Value as PG's `NULL` value.

(Still do not understand the problem, it seems.)

**Comment by tchaloupka** [comment]
:)
Ok I think that we can settle on that empty or null string written to DB is resulting to empty string column.

But what happens here?
```D
auto v = Value(ValueFormat.BINARY, OidType.Text);
assert(v.isNull); //ok
assert(v.as!string == ''); // ok null string and '' both pass
assert(v.as!string is null); // this is the question - should pass or not?
```

**Comment by denizzzka** [comment]
> assert(v.as!string == ''); // ok null string and '' both pass

This is a bug here:
https://github.com/denizzzka/dpq2/blame/master/src/dpq2/conv/to_d_types.d#L100

**Comment by denizzzka** [comment]
https://github.com/denizzzka/dpq2/tree/nullable_fix

My NULL issue fix

**Comment by tchaloupka** [comment]
Oh so then exception should be thrown then.

To make sure before I'll make the changes :)

Writing to the DB:

string value| resulting DB value
---------------|-----------------------
`null`|`''`
`""`|`''`
`Nullable!string.init`|`NULL`
`Nullable!string("")`|`''`

Reading from the DB:

DB value|D type|result
-|-|-
`NULL`|`v.as!string`|throws exception
`NULL`|`v.as!(Nullable!string)`|`Nullable!string.init`
`''`|`v.as!string`|`""`
`''`|`v.as!(Nullable!string)`|`Nullable!string("")`

**Comment by denizzzka** [comment]
All tables are correct. This behaviour is implemented in #96

File: src/dpq2/conv/to_d_types.d [hunk 1]
```diff
@@ -97,7 +106,10 @@ auto tunnelForBinaryValueAsCalls(T)(in Value v)
 
 string valueAsString(in Value v) pure
 {
-    if (v.isNull) return null;
+    import core.exception: AssertError;
+    import std.exception: enforceEx;
+
+    enforceEx!AssertError(!v.isNull, "Attempt to read NULL value", __FILE__, __LINE__);
```
Comment (denizzzka) [inline src/dpq2/conv/to_d_types.d]: already checked in Value.data()

**Comment by denizzzka** [comment]
`rawValueToNumeric` isn't covered by Nullable stuff.

String like:
```D
C!(Nullable!StringFoo)(Nullable!StringFoo(StringFoo.baz), "text", "'baz'");
```
isn't tested.

**Comment by tchaloupka** [comment]
Made additional changes in regard with the competing PR.

File: src/dpq2/oids.d [hunk 2]
```diff
@@ -156,15 +156,18 @@ OidType detectOidTypeFromNative(T)()
...
+    import std.traits : isIntegral, OriginalType, TemplateArgsOf, Unqual;
     import std.typecons : Nullable;
     static import dpq2.conv.time;
     import vibe.data.json : VibeJson = Json;
 
     static if(is(Unqual!T == Nullable!R,R))
         return detectOidTypeFromNative!(TemplateArgsOf!T[0]); //remove Nullable from Type before test
     else static if (is(Unqual!T == enum))
+    {
+        static assert(isIntegral!T || is(OriginalType!T == string), "Only integral and string enums are supported");
```
Comment (denizzzka) [inline src/dpq2/oids.d]: This brings us back to the problem of `long` type support.

Can you add check of support long enums sized more than `int.max` into `~master`? Then we can simple merge it into one of theese PRs.

Comment (denizzzka) [reply src/dpq2/oids.d]: see #97

**Comment by denizzzka** [comment]
So, our patches are almost identical, almost do not differ?

**Comment by tchaloupka** [comment]
Yep, basically almost the same except for some style difference and not sorted out enums handling.
Mine also has some documentation of the behaviour added and added `name` PG type to be accepted as string. But that can be added after merge in separate PR.

**Comment by denizzzka** [comment]
> But that can be added after merge in separate PR.

That would be good.

**Comment by denizzzka** [comment]
Shit! Accidentally merged this PR! Reverting

**Pull request merged**